backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/scope/model/impl/ElasticsearchScopeModel.java

```diff
@@ -130,15 +130,32 @@ public EventContext getIndexesEventContext() {
 	}
 
 	public boolean hasSchemaObjectNodeComponent(String absoluteFieldPath) {
+		Boolean result = null;
+
+		// storing these variables to build the exception in case of incompatibility
+		String selectedIndexName = null;
+		ElasticsearchIndexSchemaObjectNode selectedObjectNode = null;
+
 		for ( ElasticsearchIndexModel indexModel : indexModels ) {
+			String indexName = indexModel.getHibernateSearchIndexName();
+
 			ElasticsearchIndexSchemaObjectNode objectNode = indexModel.getObjectNode( absoluteFieldPath );
-			// TODO HSEARCH-2389 check multi-indexes model incompatibility
-			if ( objectNode != null ) {
-				return true;
+			Boolean hasObjectNode = ( objectNode != null );
+
+			if ( result == null ) {
+				result = hasObjectNode;
+				selectedIndexName = indexName;
+				selectedObjectNode = objectNode;
+			}
+			else if ( !result.equals( hasObjectNode ) ) {
```

Comment:
So you're throwing an exception in the following cases:

1. index A has an object field, index B has a non-object field
2. index A has an object field, index B doesn't have any field with this name (object or non-object).

I agree with 1, but 2 should not throw an exception.

You probably need more tests?